swarmrl/engine/gaurav_sim.py

```diff
@@ -0,0 +1,697 @@
+        dynamic_viscosity=params.dynamic_viscosity.m_as("sim_dyn_viscosity"),
+        fluid_density=params.fluid_density.m_as("sim_mass / sim_length**3"),
+        lubrication_threshold=params.lubrication_threshold.m_as("sim_length"),
+        magnetic_constant=params.magnetic_constant.m_as("sim_magnetic_permeability"),
+        capillary_force_data_path=params.capillary_force_data_path,
+    )
+    return params_simunits
+
+
+@dataclasses.dataclass
+class Raft:
+    pos: np.array
+    alpha: float  # in radians
+    magnetic_moment: float
+
+    def get_director(self) -> np.array:
+        return np.array([np.cos(self.alpha), np.sin(self.alpha)])
+
+
+@dataclasses.dataclass
+class GauravAction:
+    amplitudes: np.array
+    frequencies: np.array
+    phases: np.array
+    offsets: np.array
+
+
+class Model:
+    def calc_action(self, rafts: typing.List[Raft]) -> GauravAction:
+        raise NotImplementedError
+
+
+class ConstantAction(Model):
+    def __init__(self, action: GauravAction) -> None:
+        self.action = action
+
+    def calc_action(self, rafts: typing.List[Raft]) -> GauravAction:
+        return self.action
+
+
+def calc_B_field(action: GauravAction, t: float):
+    angles = 2 * np.pi * action.frequencies * t + action.phases
+    return action.amplitudes * np.sin(angles) + action.offsets
+
+
+@numba.jit
+def calc_f_dipole(
+    r_vec: np.array,
+    phi_1: float,
+    phi_2: float,
+    m_1: float,
+    m_2: float,
+    mag_const: float,
+):
+    """
+    eq. 42, "AN ANALYTIC SOLUTION FOR THE FORCE BETWEEN TWO MAGNETIC DIPOLES"
+    """
+    m1_vec = m_1 * np.array([np.cos(phi_1), np.sin(phi_1), 0])
+    m2_vec = m_2 * np.array([np.cos(phi_2), np.sin(phi_2), 0])
+    r = np.linalg.norm(r_vec)
+    r_hat = np.zeros(3)
+    r_hat[0] = r_vec[0] / r
+    r_hat[1] = r_vec[1] / r
+    f = (
+        3
+        * mag_const
+        / (4 * np.pi * r**4)
+        * (
+            np.cross(np.cross(r_hat, m1_vec), m2_vec)
+            + np.cross(np.cross(r_hat, m2_vec), m1_vec)
+            - 2 * np.dot(m1_vec, m2_vec) * r_hat
+            + 5 * r_hat * np.dot(np.cross(r_hat, m1_vec), np.cross(r_hat, m2_vec))
+        )
+    )
+    return f[:2]
+
+
+@numba.jit
+def calc_lubG(x, radius):
+    if x == 0:
+        return 0.0212758 / radius**3
+    else:
+        logx = np.log(x)
+        return ((0.0212758 * (-logx) + 0.181089) * (-logx) + 0.381213) / (
+            radius**3 * ((-logx) * ((-logx) + 6.0425) + 6.32549)
+        )
+
+
+@numba.jit
+def calc_lubA(x, radius):
+    if x == 0:
+        return 0
+    else:
+        return x * (-0.285524 * x + 0.095493 * x * np.log(x) + 0.106103) / radius
+
+
+@numba.jit
+def calc_lubB(x: float, radius: float):
+    if x == 0.0:
+        return 0.0212764 / radius
+    else:
+        logx = np.log(x)
+        return ((0.0212764 * (-logx) + 0.157378) * (-logx) + 0.269886) / (
+            radius * (-logx) * ((-logx) + 6.0425) + 6.32549
+        )
+
+
+@numba.jit
```

Comment:
Is there a reason to with numba here and not Jax? If one requirement already includes GPU support and JIT compile I would lean towards sticking with that unless there is a reason not to.